tutorial/tests/test_functions.py

```diff
@@ -69,97 +58,172 @@ def test_greet(
+    return_annotation = signature.return_annotation
+
+    assert function_to_test.__doc__ is not None, "The function is missing a docstring."
+    assert (
+        len(params) == 2
+    ), "The function should take exactly two arguments (length and width)."
+    assert all(
+        p in params.keys() for p in ["length", "width"]
+    ), "The function's parameters should be 'length' and 'width'."
+    assert all(
+        p.annotation is float for p in params.values()
+    ), "Both parameters should be annotated as float."
+    assert return_annotation is str, "The return type should be annotated as str."
+
+
+@pytest.mark.parametrize(
+    "length,width",
+    [
+        (2.0, 3.0),
+        (5.0, 4.0),
+        (1.5, 2.5),
+        (0.1, 0.1),
+    ],
+)
+def test_calculate_basic_area(length: float, width: float, function_to_test):
+    validate_basic_area_signature(function_to_test)
+    expected = reference_calculate_basic_area(length, width)
+    result = function_to_test(length, width)
+    assert isinstance(result, str), "Result should be a string"
+    assert expected == result, "Incorrect area calculation or formatting"
+
+
+# Part 2
+
+
+def reference_calculate_metric_area(
+    length: float, width: float, unit: str = "cm"
+) -> str:
+    """Reference solution for Part 2: metric units only."""
+    if unit not in ("cm", "m"):
+        return f"Invalid unit: {unit}"
+
+    if unit == "m":
+        length *= 100
+        width *= 100
+
+    area = length * width
+    return f"{area:.2f} cm^2"
+
+
+def validate_metric_area_signature(function_to_test) -> None:
+    """Validate signature of the metric area calculation function."""
+    signature = inspect.signature(function_to_test)
+    params = signature.parameters
+    return_annotation = signature.return_annotation
+
+    assert function_to_test.__doc__ is not None, "The function is missing a docstring."
+    assert (
+        len(params) == 3
+    ), "The function should take three arguments (length, width, and unit)."
+    assert all(
+        p in params.keys() for p in ["length", "width", "unit"]
+    ), "The function's parameters should be 'length', 'width' and 'unit'."
+    assert (
+        params["length"].annotation is float
+    ), "Parameter 'length' should be annotated as float."
+    assert (
+        params["width"].annotation is float
+    ), "Parameter 'width' should be annotated as float."
+    assert (
+        params["unit"].annotation is str
+    ), "Parameter 'unit' should be annotated as str."
+    assert (
+        params["unit"].default == "cm"
+    ), "Parameter 'unit' should have a default value of 'cm'."
+    assert return_annotation is str, "The return type should be annotated as str."
+
+
+@pytest.mark.parametrize(
+    "length,width,unit",
+    [
+        (2.0, 3.0, "cm"),
+        (2.0, 3.0, "m"),
+        (1.5, 2.0, "cm"),
+        (1.5, 2.0, "m"),
+    ],
+)
+def test_calculate_metric_area(length, width, unit, expected, function_to_test):
```

Comment:
```suggestion
def test_calculate_metric_area(length, width, unit, function_to_test):
```

This test didn't compile for me, as far as i can tell the expected is not given here so should be removed

Reply:
That's correct, I'll fix it.